src/time.rs

```diff
@@ -0,0 +1,189 @@
+/*! Time structures.
+The `time` modules contains structures used to represent both
+absolute and relative time.
+*/
+
+use core::{ops, fmt};
+
+const MILLIS_PER_SEC: i64 = 1_000;
+
+/// A representation of an absolute time value.
+///
+/// The `Instant` type is a wrapper around a `i64` value that
+/// represents a number of milliseconds, monotonically increasing
+/// since an arbitrary moment in time, such as system startup.
+///
+/// * A value of `0` is inherently arbitrary.
+/// * A value less than `0` indicates a time before the starting
+///   point.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Instant {
+    millis: i64,
+}
+
+impl Instant {
+    /// Create a new `Instant` from a number of milliseconds.
+    pub fn from_millis(millis: i64) -> Instant {
+        Instant { millis }
+    }
+
+    /// The fractional number of milliseconds that have passed
+    /// since the beginning of time.
+    pub fn millis(&self) -> i64 {
+        self.millis % MILLIS_PER_SEC
+    }
+
+    /// The number of whole seconds that have passed since the
+    /// beginning of time.
+    pub fn secs(&self) -> i64 {
+        self.millis / MILLIS_PER_SEC
+    }
+
+    /// The total number of milliseconds that have passed since
+    /// the biginning of time.
+    pub fn total_millis(&self) -> i64 {
+        self.millis
+    }
+}
+
+impl fmt::Display for Instant {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}.{}s", self.secs(), self.millis())
+    }
+}
+
+impl ops::Add<Duration> for Instant {
+    type Output = Instant;
+
+    fn add(self, rhs: Duration) -> Instant {
+        Instant::from_millis(self.millis + rhs.total_millis())
+    }
+}
+
+impl ops::Sub<Duration> for Instant {
+    type Output = Instant;
+
+    fn sub(self, rhs: Duration) -> Instant {
+        Instant::from_millis(self.millis - rhs.total_millis())
+    }
+}
+
+/// A relative amount of time.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Duration {
+    millis: i64,
```

Comment:
No, _this_ should be an unsigned value. Consider a function `fn sleep(Duration)`; how can it possibly sleep for, say, minus one minute?